File: app/src/main/java/com/chris/eban/domain/usecase/EventListQuery.java
package com.chris.eban.domain.usecase;

import com.chris.eban.domain.EventListRepository;
import com.chris.eban.domain.JobThread;
import com.chris.eban.domain.SingleUseCase;
import com.chris.eban.domain.entity.DMEventListItem;

import java.util.List;

import io.reactivex.Single;
import io.reactivex.functions.Function;

public class EventListQuery extends SingleUseCase<List<DMEventListItem>> {

    private JobThread jobThread;
    private EventListRepository repository;

    public EventListQuery(JobThread jobThread, EventListRepository repository) {
        this.jobThread = jobThread;
        this.repository = repository;
    }

    @Override
    protected Single<List<DMEventListItem>> buildSingle() {
        return Single.just(repository)
                .map(new Function<EventListRepository, List<DMEventListItem>>() {
                    @Override
                    public List<DMEventListItem> apply(EventListRepository repository) {
                        return repository.queryEventList();
                    }
                })
                .subscribeOn(jobThread.provideWorker())
                .observeOn(jobThread.provideUI());
    }
}

File: app/src/main/java/com/chris/eban/domain/usecase/EventListQuery.kt
package com.chris.eban.domain.usecase

import com.chris.eban.domain.EventListRepository
import com.chris.eban.domain.JobThread
import com.chris.eban.domain.SingleUseCase
import com.chris.eban.domain.entity.DMEventListItem
import io.reactivex.Single

class EventListQuery(private val jobThread: JobThread,
                     private val repository: EventListRepository)
    : SingleUseCase<List<DMEventListItem>>() {

    override fun buildSingle(): Single<List<DMEventListItem>>? {
        return Single.just(repository)
                .map { repository -> repository.queryEventList() }
                .subscribeOn(jobThread.provideWorker())
                .observeOn(jobThread.provideUI())
    }
}
